src/ess/reduce/time_of_flight/toa_to_tof.py

```diff
@@ -0,0 +1,520 @@
+    return UnwrappedTimeOfArrivalMinusPivotTime(
+        -pivot_time.to(unit=elem_unit(toa), copy=False) + toa
+    )
+
+
+def time_of_arrival_minus_pivot_time_modulo_period(
+    toa_minus_pivot_time: UnwrappedTimeOfArrivalMinusPivotTime,
+    frame_period: FramePeriod,
+) -> TimeOfArrivalMinusPivotTimeModuloPeriod:
+    """
+    Compute the time of arrival of the neutron at the detector, unwrapped at the pulse
+    period, minus the start time of the frame, modulo the frame period.
+
+    Parameters
+    ----------
+    toa_minus_pivot_time:
+        Time of arrival of the neutron at the detector, unwrapped at the pulse period,
+        minus the start time of the frame.
+    frame_period:
+        Period of the frame, i.e., time between the start of two consecutive frames.
+    """
+    return TimeOfArrivalMinusPivotTimeModuloPeriod(
+        toa_minus_pivot_time
+        % frame_period.to(unit=elem_unit(toa_minus_pivot_time), copy=False)
+    )
+
+
+def time_of_arrival_folded_by_frame(
+    toa: TimeOfArrivalMinusPivotTimeModuloPeriod,
+    pivot_time: PivotTimeAtDetector,
+) -> FrameFoldedTimeOfArrival:
+    """
+    The time of arrival of the neutron at the detector, folded by the frame period.
+
+    Parameters
+    ----------
+    toa:
+        Time of arrival of the neutron at the detector, unwrapped at the pulse period,
+        minus the start time of the frame, modulo the frame period.
+    pivot_time:
+        Pivot time at the detector, i.e., the time of the start of the frame at the
+        detector.
+    """
+    return FrameFoldedTimeOfArrival(
+        toa + pivot_time.to(unit=elem_unit(toa), copy=False)
+    )
+
+
+def time_of_flight_data(
+    da: RawData,
+    lookup: MaskedTimeOfFlightLookupTable,
+    ltotal: Ltotal,
+    toas: FrameFoldedTimeOfArrival,
+) -> TofData:
+    from scipy.interpolate import RegularGridInterpolator
+
+    # TODO: to make use of multi-threading, we could write our own interpolator.
+    # This should be simple enough as we are making the bins linspace, so computing
+    # bin indices is fast.
+    f = RegularGridInterpolator(
+        (
+            lookup.coords["toa"].to(unit=elem_unit(toas), copy=False).values,
+            lookup.coords["distance"].to(unit=ltotal.unit, copy=False).values,
+        ),
+        lookup.data.to(unit=elem_unit(toas), copy=False).values.T,
+        method="linear",
+        bounds_error=False,
+    )
+
+    if da.bins is not None:
+        ltotal = sc.bins_like(toas, ltotal).bins.constituents["data"]
+        toas = toas.bins.constituents["data"]
+
+    tofs = sc.array(
+        dims=toas.dims, values=f((toas.values, ltotal.values)), unit=elem_unit(toas)
+    )
+
+    out = da.copy(deep=False)
+    if out.bins is not None:
+        parts = out.bins.constituents
+        out.data = sc.bins(**parts)
+        parts["data"] = tofs
+        out.bins.coords["tof"] = _bins_no_validate(**parts)
+    else:
+        out.coords["tof"] = tofs
```

Comment:
Can use `da.bins.assign_coords` and `da.assign_coords` to avoid some back and forth?

Reply:
I changed it to
```Py
    if da.bins is not None:
        parts = da.bins.constituents
        parts["data"] = tofs
        out = da.bins.assign_coords(tof=_bins_no_validate(**parts))
    else:
        out = da.assign_coords(tof=tofs)
```
Seems to work. Is that what you meant?

Reply:
Yes, `parts` is just a `dict`.

Reply:
Hmm, I need `scipp=25.01` to use `da.bins.assign_coords` :-(

Reply:
Seems it was released now.